NativeImageTest that uses @Inject: better error message than NullPointerException

Given a test that runs

```
@QuarkusTest
public class TimeTableResourceTest {

    @Inject
    private TimeTableResource timeTableResource;

    @Test
    public void solveDemoDataUntilFeasible() {
        timeTableResource.solve();
        ...
   }
}
```

The native test version of it:
```

@NativeImageTest
public class TimeTableResourceIT extends TimeTableResourceTest {...}
```

fails because `@Inject` is not supported in native tests:

```
[INFO] Running org.acme.optaplanner.rest.TimeTableResourceIT
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.014 s <<< FAILURE! - in org.acme.optaplanner.rest.TimeTableResourceIT
[ERROR] solveDemoDataUntilFeasible  Time elapsed: 0.013 s  <<< ERROR!
java.lang.NullPointerException
```

**The error message `java.lang.NullPointerException` is not helpful to determine the cause. Can we change that to something like `Don't use @Inject in a native test.`?**

--- a/integration-tests/kafka-streams/src/test/java/io/quarkus/it/kafka/streams/KafkaStreamsTest.java
+++ b/integration-tests/kafka-streams/src/test/java/io/quarkus/it/kafka/streams/KafkaStreamsTest.java
@@ -6,8 +6,6 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
 
-import javax.inject.Inject;
-
 import org.apache.http.HttpStatus;
 import org.apache.kafka.clients.consumer.Consumer;
 import org.apache.kafka.clients.consumer.ConsumerConfig;
@@ -20,7 +18,6 @@ import org.apache.kafka.clients.producer.ProducerConfig;
 import org.apache.kafka.clients.producer.ProducerRecord;
 import org.apache.kafka.common.serialization.IntegerDeserializer;
 import org.apache.kafka.common.serialization.IntegerSerializer;
-import org.eclipse.microprofile.config.Config;
 import org.hamcrest.CoreMatchers;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
@@ -34,9 +31,6 @@ import io.restassured.RestAssured;
 @QuarkusTestResource(KafkaTestResource.class)
 @QuarkusTest
 public class KafkaStreamsTest {
-
-    @Inject
-    public Config config;
 
     private static Producer<Integer, Customer> createCustomerProducer() {
         Properties props = new Properties();

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/NativeImageTest.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/NativeImageTest.java
@@ -20,6 +20,9 @@ import org.junit.jupiter.api.extension.ExtendWith;
  * run, it is expected that the JVM tests will be standard unit tests that are
  * executed by surefire, while the native image tests will be integration tests
  * executed by failsafe.
+ * This also means that injection of beans into a test class using {@code @Inject} is not supported
+ * in native image tests. Such injection is only possible in tests injected with
+ * {@link @QuarkusTest} so the test class structure must take this into account.
  *
  */
 @Target(ElementType.TYPE)

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/NativeTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/NativeTestExtension.java
@@ -2,7 +2,10 @@ package io.quarkus.test.junit;
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.lang.reflect.Field;
 import java.util.Map;
+
+import javax.inject.Inject;
 
 import org.junit.jupiter.api.extension.AfterEachCallback;
 import org.junit.jupiter.api.extension.BeforeAllCallback;
@@ -41,6 +44,7 @@ public class NativeTestExtension
 
     @Override
     public void beforeAll(ExtensionContext extensionContext) throws Exception {
+        ensureNoInjectAnnotationIsUsed(extensionContext.getRequiredTestClass());
 
         ExtensionContext root = extensionContext.getRoot();
         ExtensionContext.Store store = root.getStore(ExtensionContext.Namespace.GLOBAL);
@@ -69,6 +73,23 @@ public class NativeTestExtension
                 throw new JUnitException("Quarkus native image start failed, original cause: " + e);
             }
         }
+    }
+
+    private void ensureNoInjectAnnotationIsUsed(Class<?> testClass) {
+        Class<?> current = testClass;
+        while (current.getSuperclass() != null) {
+            for (Field field : current.getDeclaredFields()) {
+                Inject injectAnnotation = field.getAnnotation(Inject.class);
+                if (injectAnnotation != null) {
+                    throw new JUnitException(
+                            "@Inject is not supported in NativeImageTest tests. Offending field is "
+                                    + field.getDeclaringClass().getTypeName() + "."
+                                    + field.getName());
+                }
+            }
+            current = current.getSuperclass();
+        }
+
     }
 
     @Override

